### PR TITLE
Dynamically deserialize networktables structs

### PR DESCRIPTION
**Pull request by onlycs**
image
image

Very simply, we get struct information by listening for all announced topics and if it is has type `structschema`, we store it and parse as-needed into a `DynStructSchema` (`lib/services/struct_schemas/dyn_struct.dart`). When we subscribe to or get data from a topic that uses a schema, we provide `DynStructSchema` along with the `Uint8List` to `DynStruct` which recursively parses struct fields and values.

I also made saving actually save to a file

### CODE REVIEW

**Comment by Gold872** [comment]
Leaving this comment here to let you know that this is on my list of things to review, it's just a lot of changes and will take a bit of time to go through, I should have some time next week

Excellent work though, I'm very impressed with the implementation!

File: lib/pages/dashboard_page.dart [hunk 1]
```diff
@@ -599,29 +593,30 @@ class _DashboardPageState extends State<DashboardPage> with WindowListener {
...
     } catch (e) {
       _showJsonLoadingError(e.toString());
       return;
     }
 
-    await preferences.setString(PrefKeys.layout, jsonEncode(jsonData));
+    await preferences.setString(PrefKeys.layoutPath, file.path);
 
     setState(() => _loadLayoutFromJsonData(jsonString));
   }
```
Comment (Gold872) [inline lib/pages/dashboard_page.dart:606]: The layout is stored in Shared Preferences to avoid an issue where deleting the file would delete the layout too. It could make sense to have the last path as an "initial directory" for the user, but when importing a layout I don't think it would be expected for that to turn into the layout location.

File: lib/services/log.dart [hunk 2]
```diff
@@ -34,13 +34,12 @@ class Log {
         if (kReleaseMode) FileOutput(file: logFile),
       ]),
       filter: ProductionFilter(),
+      level: Level.debug,
```
Comment (Gold872) [inline lib/services/log.dart:37]: Make sure to remove this at some point so it doesn't mess with the global log level

File: lib/services/log.dart [hunk 3]
```diff
@@ -34,13 +34,12 @@ class Log {
...
     );
   }
 
   void log(Level level, dynamic message, [dynamic error, StackTrace? trace]) {
-    if (Logger.level.value > level.value) {
-      return;
-    }
+    if (Logger.level.value > level.value) {}
```
Comment (Gold872) [inline lib/services/log.dart:42]: Same as above

File: lib/services/nt_connection.dart [hunk 4]
```diff
@@ -63,6 +66,18 @@ class NTConnection {
       topic: '',
       options: const NT4SubscriptionOptions(topicsOnly: true),
     );
+
+    _ntClient.addTopicAnnounceListener((NT4Topic topic) {
+      if (topic.type == "structschema") {
+        subscribe(topic.name).listen((data, _) {
+          if (data == null) return;
+          String schema = String.fromCharCodes(data as List<int>);
```
Comment (Gold872) [inline lib/services/nt_connection.dart:74]: I'd prefer having some sort of type checking, since otherwise what could happen is the app suddenly breaks without any notification to the user

File: lib/services/nt_widget_builder.dart [hunk 5]
```diff
@@ -49,13 +50,15 @@ typedef NTModelJsonProvider = NTWidgetModel Function({
...
   required SharedPreferences preferences,
+  required NT4StructMeta? ntStructMeta,
 });
 
 typedef NTModelProvider = NTWidgetModel Function({
   String dataType,
   required NTConnection ntConnection,
   double period,
   required SharedPreferences preferences,
+  required NT4StructMeta? ntStructMeta,
```
Comment (Gold872) [inline lib/services/nt_widget_builder.dart:61]: I don't think a nullable field should be required, since for models that don't depend on a struct subscription (any multi-topic widget), it doesn't entirely make sense for it to have to be given a struct meta

File: lib/services/shuffleboard_nt_listener.dart [hunk 6]
```diff
@@ -576,25 +576,21 @@ class ShuffleboardNTListener {
...
         }
       } else {
         if (shuffleboardTreeRoot.hasRow(row)) {
           current = shuffleboardTreeRoot.getRow(row);
         } else {
           current = shuffleboardTreeRoot.createNewRow(
-              topic: currentTopic,
-              name: row,
-              ntTopic: (lastElement) ? nt4Topic : null);
+              topic: currentTopic, name: row, entry: null);
```
Comment (Gold872) [inline lib/services/shuffleboard_nt_listener.dart:593]: There isn't any issue here but I do plan on removing the Shuffleboard API support for 2026, so these changes won't be needed

File: lib/services/struct_schemas/dyn_struct.dart [hunk 7]
```diff
@@ -0,0 +1,370 @@
+import 'dart:typed_data';
+
+import 'package:flutter/foundation.dart';
+
+import 'package:dot_cast/dot_cast.dart';
+
+class DynStructField {
```
Comment (Gold872) [inline lib/services/struct_schemas/dyn_struct.dart:7]: I prefer keeping class names less truncated, maybe rename this to `DynamicStructField`?

File: lib/services/struct_schemas/dyn_struct.dart [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+  bool get boolValue => (this as DynStructBoolean).value;
+  int get intValue => (this as DynStructInt).value;
+  int get longValue => (this as DynStructLong).value;
+  double get floatValue => (this as DynStructFloat).value;
+  double get doubleValue => (this as DynStructDouble).value;
+  String get stringValue => (this as DynStructString).value;
+  DynStructValue? get nullableValue => (this as DynStructNullable).value;
+  List<DynStructValue> get arrayValue => (this as DynStructArray).value;
+  DynStruct get structValue => (this as DynStructStruct).value;
+}
```
Comment (Gold872) [inline lib/services/struct_schemas/dyn_struct.dart:194]: It might be possible to have this be a generic class (like DynamicStructValue<T>) instead of having to cast, and the `value` field can be of type `T?`

**Review by Gold872 — CHANGES_REQUESTED** [review]
Sorry for taking so long to get to this!

I'm very impressed that you were able to get this implemented! In the past I've avoided adding struct decoding since it felt out of scope with Elastic, but with more recent (and planned) struct integration throughout WPILib it would make sense to have the ability to display (but not edit) individual struct fields.

That being said though, there's still some things I would like to have changed/fixed. Firstly, I'm a bit confused between the difference between `NT4StructMeta` and `DynStructSchema`. I'm also not sure if it entirely makes sense to have the widgets store the entire schema in its json, since it means that any time a struct field widget is created, it will assume that the schema has never changed. For example, if a widget is created with a struct schema for some type of Swerve State, but then the dashboard connects to a robot with a different struct that has the same name, the widget will always assume that it's the same which can cause invalid values to be displayed. It might make sense to instead store the name of the struct and its field, and have it looked up whenever a new value comes in.

Lastly, I would want to have unit tests for the schema parsing, decoding, and displaying. Right now that might not be easy to add since there's a lot of stuff still being changed around, but I would definitely want lots of these new features in unit tests before being merged. They can be a pain to write, but they've saved me from breaking a lot of important features.

File: lib/pages/dashboard_page.dart (hunk 1 above)
Comment (onlycs) [reply lib/pages/dashboard_page.dart:606]: Yeah I made this change cuz I was kind of annoyed when I clicked save and elastic didn't save to the file. I can make this save to user prefs as well and load it when the json is deleted.

File: lib/services/nt_connection.dart (hunk 4 above)
Comment (onlycs) [reply lib/services/nt_connection.dart:74]: Type checking as in? That's kind of the only way (that I know how, at least) to check whether an NT field is a struct schema... Struct values are prefixed with `struct:`.

Probably misunderstood you, not really sure what you mean.

File: lib/services/struct_schemas/dyn_struct.dart (hunk 8 above)
Comment (onlycs) [reply lib/services/struct_schemas/dyn_struct.dart:194]: I come from rust and was trying to use an enum lol

There's also the problem that NT fields can only be of a certain few types. Giving T the possibility to be any type, at least in my opinion, is probably not the best, unless it can be constrained in some way I don't know about

File: lib/services/nt_connection.dart (hunk 4 above)
Comment (onlycs) [reply lib/services/nt_connection.dart:74]: Also wouldn't the app break if we suddenly switched some published NT field from a string to a double between launches?

Also, I'm like 80% sure the code as it stands will bail out by displaying the raw bytes if the schema suddenly changes like that

**Comment by onlycs** [comment]
Thanks for reviewing it and sorry for the late reply. Comps keep me busy!

DynStructSchema is a data class that contains the schema of the struct (i.e. fields and field names) NT4StructMeta contains a NT path to the root value (containing the entire struct and all of its data), a path to the subvalue (e.g. `pose.rotation.value` to get a Pose2d's angle in radians), and the schema of the struct it's using to parse. Basically, schema is for the struct as a whole and meta is for the value field in the struct which is used by elastic

I'm 80% sure the code as it stands will bail out by showing raw bytes if the schemas don't line up for some reason.

File: lib/services/nt_widget_builder.dart (hunk 5 above)
Comment (onlycs) [reply lib/services/nt_widget_builder.dart:61]: I'm not sure how to do this without the required field, the register() function would need an overhaul, probably with a seperate type for nt multitopic model providers.

I'll begin work on this

**Comment by onlycs** [comment]
I went ahead and removed ntStructMeta from the multitopic widgets

**Comment by onlycs** [comment]
i was in the middle of adding typechecking to nt types so I committed those changes too. im going to mark this as a draft until i finish fixing tests

**Comment by onlycs** [comment]
Also there's a problem where widgets in list layouts are too small. Dunno why that's happening; afaik I didn't change anything relating to that

File: lib/services/nt_connection.dart [hunk 9]
```diff
@@ -36,6 +37,8 @@ class NTConnection {
   Map<int, NT4Subscription> subscriptionMap = {};
   Map<NT4Subscription, int> subscriptionUseCount = {};
 
+  Map<String, String> knownSchemas = {};
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/services/nt_connection.dart:40]: Thats probably in there from before and unused, will remove

File: lib/widgets/nt_widgets/multi_topic/field_widget.dart [hunk 10]
```diff
@@ -536,7 +535,7 @@ class FieldWidget extends NTWidget {
         minHeight: 4.0,
       ),
       decoration: BoxDecoration(
-        color: Colors.black.withValues(alpha: 0.35),
+        color: Colors.black.withAlpha(90),
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/widgets/nt_widgets/multi_topic/field_widget.dart:538]: It errored for me when I tried to compile, idk why. Will take another look

File: lib/widgets/nt_widgets/single_topic/boolean_box.dart [hunk 11]
```diff
@@ -109,8 +111,8 @@ class BooleanBoxModel extends SingleTopicNTWidgetModel {
       }
     }
 
-    _trueColor = Color(trueColorValue ?? Colors.green.toARGB32());
-    _falseColor = Color(falseColorValue ?? Colors.red.toARGB32());
+    _trueColor = trueColor;
+    _falseColor = falseColor;
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/widgets/nt_widgets/single_topic/boolean_box.dart:115]: this was another thing giving me some errors when i tried to compile once, ill take another look

File: lib/widgets/draggable_dialog.dart [hunk 12]
```diff
@@ -34,12 +31,12 @@ class _DraggableDialogState extends State<DraggableDialog> {
...
-          .toSet(),
-      supportedResizeDevices: PointerDeviceKind.values
-          .whereNot((e) => e == PointerDeviceKind.trackpad)
-          .toSet(),
+      // supportedDragDevices: PointerDeviceKind.values
+      //     .whereNot((e) => e == PointerDeviceKind.trackpad)
+      //     .toSet(),
+      // supportedResizeDevices: PointerDeviceKind.values
+      //     .whereNot((e) => e == PointerDeviceKind.trackpad)
+      //     .toSet(),
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/widgets/draggable_dialog.dart:39]: I thought I had commented this out for some reason when hastily hacking together the first few commits. Ill undo.

File: lib/services/nt_connection.dart [hunk 13]
```diff
@@ -63,6 +66,9 @@ class NTConnection {
       topic: '',
       options: const NT4SubscriptionOptions(topicsOnly: true),
     );
+
+    // add all struct schemas to the schema manager
+    SchemaInfo.getInstance().listen(_ntClient);
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/services/nt_connection.dart:71]: `SchemaInfo` is used all over the place... passing it around where ever it's needed gets annoying, fast. I don't think its the best idea to have this be a field in `NTConnection` unless it's static and globally accessible.

File: lib/widgets/nt_widgets/nt_widget.dart [hunk 14]
```diff
@@ -139,80 +142,86 @@ class SingleTopicNTWidgetModel extends NTWidgetModel {
...
+    required ntStructMeta,
+    this.dataType,
     super.period,
   })  : _typeOverride = type,
         super();
 
   SingleTopicNTWidgetModel.fromJson({
     required super.ntConnection,
     required super.preferences,
+    required ntStructMeta,
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/widgets/nt_widgets/nt_widget.dart:154]: Shouldn't it be `this.ntStructMeta`... it lives in SingleTopic

File: lib/widgets/network_tree/networktables_tree.dart [hunk 15]
```diff
@@ -58,6 +60,7 @@ class _NetworkTableTreeState extends State<NetworkTableTree> {
     rowName: '',
   );
   late final TreeController<NetworkTableTreeRow> treeController;
+  late final HashMap<String, void> topics = HashMap<String, void>();
```
[the remark this answers is not in the document]
Comment (onlycs) [reply lib/widgets/network_tree/networktables_tree.dart:63]: it turns out, so am I. that map was not used.

**Review by onlycs — COMMENTED** [review]
@Gold872 I made most of the changes, see my comments for more info

File: lib/services/nt_connection.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gold872) [reply lib/services/nt_connection.dart:70]: If we have it as a field in NTConnection, it will be accessible by everything that has access to it, and all parts of the program that access network tables already have access to an instance so adding it as a field shouldn't cause too many issues

[the remark this answers is not in the document]
Comment (onlycs) [reply lib/services/nt_connection.dart:70]: Specifically for `nt4_type` and `nt_struct`, it doesn't have access to the connection (it's completely independent) but it still needs access to known struct metadata for type checking and the like.

**Review by Gold872 — CHANGES_REQUESTED** [review]
This is starting to look really nice! I'll do some testing with this and make some minor modifications where I see necessary, but I'm liking the implementation of this.

One concern I do have is the way it's handling subscriptions. The logic to avoid duplicate subscriptions doesn't quite work with this since the value update is handled by the struct meta, so if we have a subscription to a struct topic (which is possible for something like a Field2d), and then a subscription to a value in that struct, it would cause issues. We could have it check the struct meta to determine if 2 subscriptions are equal, but then this would cause duplicate subscriptions to be sent to the server. Not sure how to go about preventing this.

Also, I would like to have unit tests for the schema parsing and decoding of individual struct fields, since this is where there could be a lot of edge cases and logic

**Comment by Gold872** [comment]
Unit Tests Checklist:

## Backend
- [ ] Schema Parsing (simple, sub-structs, etc)
- [ ] NT4Type parsing and conversions
- [ ] NT4Type serializing
- [ ] Schema to/from json
- [ ] Schema decoding

## Front end
- [ ] Displaying fields in tree
- [ ] Struct decoding in NT subscriptions
